contracts/base/vault/SmartVault.sol

```diff
@@ -322,9 +398,39 @@ contract SmartVault is Initializable, ERC20Upgradeable, VaultStorage, Controllab
     require(totalSupply() > 0, "no shares");
     require(numberOfShares > 0, "zero amount");
 
-    userLastWithdrawTs[msg.sender] = block.timestamp;
-
     uint256 totalSupply = totalSupply();
+
+    // this logic not eligible for normal vaults
+    // lockAllowed unchangeable attribute even for proxy upgrade process
+    if (lockAllowed()) {
+      // locking logic will add a part of locked shares as rewards for this vault
+      // calculate locked amount
+      uint256 lockStart = userLockTs[msg.sender];
+      // refresh lock time
+      // if full withdraw set timer to 0
+      if (balanceOf(msg.sender) == numberOfShares) {
+        userLockTs[msg.sender] = 0;
+      } else {
+        userLockTs[msg.sender] = block.timestamp;
+      }
+      if (lockStart != 0 && lockStart < block.timestamp) {
+        uint256 currentLockDuration = block.timestamp.sub(lockStart);
+        if (currentLockDuration < lockPeriod()) {
+          uint256 sharesBase = numberOfShares.mul(LOCK_PENALTY_DENOMINATOR - lockPenalty()).div(LOCK_PENALTY_DENOMINATOR);
+          uint256 toWithdraw = sharesBase.add(
+            numberOfShares.sub(sharesBase).mul(currentLockDuration).div(lockPeriod())
+          );
+          uint256 change = numberOfShares.sub(toWithdraw);
+          numberOfShares = toWithdraw;
+
+          _transfer(msg.sender, address(this), change);
```

Comment:
What will happen with those tokens? Could you please point me to the logic where those tokens will be used.

Reply:
it will be used as rewards